### DISCUSSION TITLE
TWIG template does not attach.

### DISCUSSION THREAD

**Discussion by kiusau** [symfony/symfony#45648] [Q&A]
Hi,  I am new to both Symfony and TWIG, and am having trouble integrating the two.  The following code successfully sends an email with a brief text message, but fails to include the HTML template.  Can you see anything missing or something that does not belong?  The attached image may be easier to read than the code appended below.

/*********************************************************************
                   spiritmail_include.html
*********************************************************************/
<?php
	ini_set('log_errors', 1);
	ini_set('error_log', dirname(__FILE__) . DIRECTORY_SEPARATOR . 'error.log');
	ini_set('html_errors', 1);
	ini_set('display_errors', 1);
	ini_set('display_startup_errors', 1);
	error_reporting(E_ALL);
	use Symfony\Component\Mailer\Mailer; 
	use Symfony\Component\Mailer\Transport;
	use Symfony\Component\Mailer\Transport\AbstractTransport;
	use Symfony\Component\Mailer\Transport\Smtp\SmtpTransport;
	use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
	use Symfony\Component\Mime;
	use Symfony\Component\Mime\Email;
	use Symfony\Bridge\Twig\Mime\TemplatedEmail;
	use Twig\Environment;
	$hostname = '...';
	$port = 587;
	$username = '...';
	$passphrase = '...';

	require_once('/home/.../vendor/autoload.php');
	$transport = new EsmtpTransport($hostname,$port,0);
	$transport->setUsername($username);
	$transport->setPassword($passphrase);

	$eMail = new TemplatedEmail();
	$eMail->from('...');
	$eMail->to($email);
	$eMail->priority(Email::PRIORITY_HIGH);
	$eMail->subject($subject);
	$eMail->text($text);
	$eMail->htmlTemplate($template);
	$eMail->context($context);

	$mailer = new Mailer($transport);
?>
/*********************************************************************
                   spiritmail.html
*********************************************************************/
<?php	
	$name = '...';
	$email = '...';
	$guid = '...';
	$text = 'My first Symfony email created with a TWIG template.';
	$template = '...';
	$context = ['name'=>$name, 'email'=>$email, 'guid'=>$guid];
	$subject = 'Confirmation Email';
	include_once ('./_utilities/php/includes/spiritmail_include.php');
	try {
		$mailer->send($eMail);
	} catch (Exception $e) {
		echo "Caught exception: " . $e->getMessage();
	}
?>
<img width="599" alt="symfony_twig_integration" src="https://user-images.githubusercontent.com/29310043/156873583-cd1e5e53-9cf8-4a1f-a902-9454c9093b68.png">

**Comment by helmis123** [comment]
Can u post a simple reproducible test ?

**Comment by kiusau** [comment]
The following link https://spiritof2021.online/rm/spiritmail.html sends an email to my local address from my remote server with good success, but the content header is plain simple text.  When the email is sent it dumps the mailer and email objects for examination.  You probably cannot reach the link as the site is in development, however.

Please find attached the message headers from the received email.

Roddy A. Stegemann, A.B., M.A., M.A.
Founder/Owner of nudge.online
www.nudge.online

***@***.***
***@***.***

Twitter: @kiusau

Mobile: +1 (206) 291-8468

909 Boren Avenue
The Cabrini, Suite 603
First Hill, Seattle
Washington, USA 98104

**Reply by kiusau** [reply]
Mailer Object
<img width="1186" alt="mailer_object" src="https://user-images.githubusercontent.com/29310043/156896887-05694b51-0a62-4d43-befc-9ee189c0bc45.png">

eMail Object
<img width="905" alt="email_object" src="https://user-images.githubusercontent.com/29310043/156896913-7df2f02e-a8de-49c0-9e3d-f822f46e6d19.png">

**Reply by kiusau** [reply]
Could it be that the $email->text() method is not compatible with the $email->htmlTemplate() method, and that the ->text() method must be replaced by the textTemplate() method? If so, how does one integrate the League converter dynamically with the formation of the email object.  As follows?  Or, does the convert() method perform the file_get_contents() and file_put_contents() for you?  Can it create the text file automatically?

******* HTML to Text Conversion *********

$htmlTemplate = /path/to/completed/template.html;
$textTemplate =/path/to/empty/template.txt;
$htmlString = file_get_contents($htmlTemplate);

use League\HTMLToMarkdown\HtmlConverter;
$converter = new HtmlConverter();
$textString = $converter->convert($htmlString);
file_put_contents($textTemplate, $textString);

******** Already formed email object **************

$email->textTemplate($textTemplate);
$email->htmlTemplate($htmlTemplate);

Please advise.

Roddy

**Reply by kiusau** [reply]
So, when I implement the above strategy.  I am told the following:

"Caught exception: A message must have a text or an HTML part or attachments."

Roddy

**Reply by kiusau** [reply]
I am going back to PHPMailer.  Sorry, Symfony.  I am out of time.
Good luck!
